src/liblelantus/lelantus_prover.cpp

```diff
@@ -101,32 +102,70 @@ void LelantusProver::generate_sigma_proofs(
-            C_.emplace_back(coin.getValue() + gs);
-
-        rA[i].randomize();
-        rB[i].randomize();
-        rC[i].randomize();
-        rD[i].randomize();
-        Tk[i].resize(params->get_sigma_m());
-        Pk[i].resize(params->get_sigma_m());
-        Yk[i].resize(params->get_sigma_m());
-        a[i].resize(params->get_sigma_n() * params->get_sigma_m());
-        sigmaProver.sigma_commit(C_, indexes[i], rA[i], rB[i], rC[i], rD[i], a[i], Tk[i], Pk[i], Yk[i], sigma[i], sigma_proofs[i]);
+
+    std::size_t threadsMaxCount = std::min((unsigned int)N, std::thread::hardware_concurrency());
+    std::vector<std::shared_future<void>> parallelTasks;
+    parallelTasks.reserve(threadsMaxCount);
+    ParallelOpThreadPool<void> threadPool(threadsMaxCount);
+
+    std::vector<std::vector<GroupElement>> C_;
+    C_.resize(N);
+    for (std::size_t j = 0; j < N; j += threadsMaxCount) {
+        for (std::size_t i = j; i < j + threadsMaxCount; ++i) {
+            if (i < N) {
+                if (!c.count(Cin[i].second))
+                    throw std::invalid_argument("No such anonymity set or id is not correct");
+
+                GroupElement gs = (params->get_g() * Cin[i].first.getSerialNumber().negate());
+                serialNumbers.emplace_back(Cin[i].first.getSerialNumber());
+
+                C_[i].reserve(c.size());
+
+                const auto& set = c.find(Cin[i].second);
+                if (set == c.end())
+                    throw std::invalid_argument("No such anonymity set");
+
+                for (auto const &coin : set->second)
+                    C_[i].emplace_back(coin.getValue() + gs);
+
+                rA[i].randomize();
+                rB[i].randomize();
+                rC[i].randomize();
+                rD[i].randomize();
+                Tk[i].resize(params->get_sigma_m());
+                Pk[i].resize(params->get_sigma_m());
+                Yk[i].resize(params->get_sigma_m());
+                a[i].resize(params->get_sigma_n() * params->get_sigma_m());
+
+                auto& sigma_i = sigma[i];
+                auto& rA_i = rA[i];
+                auto& rB_i = rB[i];
+                auto& rC_i = rC[i];
+                auto& rD_i = rD[i];
+                auto& a_i = a[i];
+                auto& Tk_i = Tk[i];
+                auto& Pk_i = Pk[i];
+                auto& Yk_i = Yk[i];
+                auto& prover = sigmaProver;
+                auto& commits = C_[i];
+                auto& index = indexes[i];
+                auto& proof = sigma_proofs[i];
+                parallelTasks.emplace_back(threadPool.PostTask([&]() {
+                    return prover.sigma_commit(commits, index, rA_i, rB_i, rC_i, rD_i, a_i, Tk_i, Pk_i, Yk_i, sigma_i, proof);
+                }));
+            } else
+                break;
+        }
+
+        for (auto& th : parallelTasks) {
+            try {
+                th.get();
+            } catch (std::exception& except) {
```

Comment:
if exception other than `std::exception` is thrown from inside `prover.sigma_commit()` it'll be rethrown on `th.get()`. Imagine situation where the first task in the list results in exception which is not supposed to abort the program. We're getting out of `LelantusProver::generate_sigma_proofs` while all other tasks are still running and referencing local variables. This will result in segfault. If `sigma_commit()` can throw you need to wrap it into catch

Reply:
also in case of client shutdown (even graceful!) all the threads are cancelled (`boost::thread_group::interrupt_all()`). In this case `th.get()` will throw boost exception, call frame will cease to exist and you'll get corrupted data or segfault